File: LeetCode/21. 合并两个有序链表.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function mergeTwoLists(list1: ListNode | null, list2: ListNode | null): ListNode | null {
    const virtualHead = new ListNode(-1);
    let p = virtualHead;
    let p1 = list1;
    let p2 = list2;
    while (p1 !== null && p2 !== null) {
        if (p1.val > p2.val) {
            p.next = p2;
            p2 = p2.next;
        } else {
            p.next = p1;
            p1 = p1.next;
        }
        p = p.next;
    }
    if (p1 !== null) p.next = p1;
    if (p2 !== null) p.next = p2;
    return virtualHead.next;
};

File: LeetCode/21. 合并两个有序链表.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
//
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn merge_two_lists(list1: Option<Box<ListNode>>, list2: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut virtual_head = ListNode::new(-1);
        let mut p = &mut virtual_head;
        let mut p1 = list1;
        let mut p2 = list2;
        while p1.is_some() && p2.is_some() {
            let (mut p1_data, mut p2_data) = (p1.as_deref_mut().unwrap(), p2.as_deref_mut().unwrap());
            if p1_data.val > p2_data.val {
                let next = p2_data.next.take();
                p.next = p2.take();
                p2 = next;
            } else {
                let next = p1_data.next.take();
                p.next = p1.take();
                p1 = next;
            }
            p = p.next.as_deref_mut().unwrap();
        }
        p.next = p1.or(p2);
        return virtual_head.next;
    }
}
